cloudnew/ec2_client.go

```diff
@@ -0,0 +1,193 @@
+
+	// RequestSpotInstances is a wrapper for ec2.RequestSpotInstances.
+	RequestSpotInstances(*ec2.RequestSpotInstancesInput) (*ec2.RequestSpotInstancesOutput, error)
+
+	// DescribeSpotInstanceRequests is a wrapper for ec2.DescribeSpotInstanceRequests.
+	DescribeSpotInstanceRequests(*ec2.DescribeSpotInstanceRequestsInput) (*ec2.DescribeSpotInstanceRequestsOutput, error)
+
+	// CancelSpotInstanceRequests is a wrapper for ec2.CancelSpotInstanceRequests.
+	CancelSpotInstanceRequests(*ec2.CancelSpotInstanceRequestsInput) error
+}
+
+// AWSClientImpl wraps ec2.EC2.
+type AWSClientImpl struct {
+	session *session.Session
+	*ec2.EC2
+}
+
+// Create a new aws-sdk-client if one does not exist, otherwise no-op.
+func (c *AWSClientImpl) Create(creds *credentials.Credentials) error {
+	if creds == nil {
+		return errors.New("creds must not be nil")
+	}
+	if c.session == nil {
+		s, err := session.NewSession()
+		if err != nil {
+			return err
+		}
+		c.session = s
+	}
+	c.EC2 = ec2.New(c.session)
+	return nil
+}
+
+// RunInstances is a wrapper for ec2.RunInstances.
+func (c *AWSClientImpl) RunInstances(input *ec2.RunInstancesInput) (*ec2.Reservation, error) {
+	return c.EC2.RunInstances(input)
+}
+
+// DescribeInstances is a wrapper for ec2.DescribeInstances
+func (c *AWSClientImpl) DescribeInstances(input *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+	return c.EC2.DescribeInstances(input)
+}
+
+// CreateTags is a wrapper for ec2.CreateTags.
+func (c *AWSClientImpl) CreateTags(input *ec2.CreateTagsInput) error {
+	_, err := c.EC2.CreateTags(input)
+	return err
+}
+
+// TerminateInstances is a wrapper for ec2.TerminateInstances.
+func (c *AWSClientImpl) TerminateInstances(input *ec2.TerminateInstancesInput) (*ec2.TerminateInstancesOutput, error) {
+	return c.EC2.TerminateInstances(input)
+}
+
+// RequestSpotInstances is a wrapper for ec2.RequestSpotInstances.
+func (c *AWSClientImpl) RequestSpotInstances(input *ec2.RequestSpotInstancesInput) (*ec2.RequestSpotInstancesOutput, error) {
+	return c.EC2.RequestSpotInstances(input)
+}
+
+// DescribeSpotInstanceRequests is a wrapper for ec2.DescribeSpotInstanceRequests.
+func (c *AWSClientImpl) DescribeSpotInstanceRequests(input *ec2.DescribeSpotInstanceRequestsInput) (*ec2.DescribeSpotInstanceRequestsOutput, error) {
+	return c.EC2.DescribeSpotInstanceRequests(input)
+}
+
+// CancelSpotInstanceRequests is a wrapper for ec2.CancelSpotInstanceRequests.
+func (c *AWSClientImpl) CancelSpotInstanceRequests(input *ec2.CancelSpotInstanceRequestsInput) error {
+	_, err := c.EC2.CancelSpotInstanceRequests(input)
+	return err
+}
+
+// AWSClientMock mocks ec2.EC2.
+type AWSClientMock struct {
+	*credentials.Credentials
+	*ec2.RunInstancesInput
+	*ec2.DescribeInstancesInput
+	*ec2.CreateTagsInput
+	*ec2.TerminateInstancesInput
+	*ec2.RequestSpotInstancesInput
+	*ec2.DescribeSpotInstanceRequestsInput
+	*ec2.CancelSpotInstanceRequestsInput
```

Comment:
does it make sense to make these anonymous? does it make things more complicated to leave them as pointers?

Reply:
I considered naming these, but leaving them like this makes it easy to find the structs you need. Pointers require that client code does the right dereferencing, or tests would fail.